mwdb/model/object.py

```diff
@@ -525,50 +525,57 @@ def _get_or_create(
         tags = tags or []
 
         is_new = False
-        new_cls = Object.get(obj.dhash).first()
+        created_object = Object.get(obj.dhash).first()
 
         # Object with the specified dhash doesn't exist - create it
-        if new_cls is None:
+        if created_object is None:
             db.session.begin_nested()
 
-            new_cls = obj
+            created_object = obj
             try:
                 # Try to create the requested object
-                new_cls.upload_time = datetime.datetime.utcnow()
-                db.session.add(new_cls)
-                db.session.flush()
+                created_object.upload_time = datetime.datetime.utcnow()
+                db.session.add(created_object)
                 db.session.commit()
                 is_new = True
             except IntegrityError:
                 # Object creation failed - probably a race condition
                 db.session.rollback()
-                new_cls = Object.get(obj.dhash).first()
-                if new_cls is None:
+                created_object = Object.get(obj.dhash).first()
+                # If object still doesn't exist, IntegrityError is not
+                # caused by concurrent upload: re-raise exception
+                if created_object is None:
                     raise
 
-        # Ensure that existing object has the expected type
-        if new_cls is not isinstance(obj, cls):
-            # If Object has been fetched, fetch typed instance
-            new_cls = cls.get(obj.dhash).first()
-            if new_cls is None:
+        if not is_new:
```

Comment:
```suggestion
        else:
```

Reply:
Not really, because `is_new` might be not set due to `IntegrityError` handling

Reply:
You are right